lib/rex/proto/ldap/server.rb

```diff
@@ -0,0 +1,298 @@
+  def dispatch_request(cli, data)
+    if self.dispatch_request_proc
+      self.dispatch_request_proc.call(cli,data)
+    else
+      default_dispatch_request(cli,data)
+    end
+  end
+
+  #
+  # Default LDAP request dispatcher
+  #
+  # @param cli [Rex::Socket::Tcp, Rex::Socket::Udp] Client sending the request
+  # @param data [String] raw LDAP request data
+  def default_dispatch_request(cli, data)
+    return if data.strip.empty?
+    data.extend(Net::BER::Extensions::String)
+    begin
+      pdu = Net::LDAP::PDU.new(data.read_ber!(Net::LDAP::AsnSyntax))
+      wlog("LDAP request has remaining data: #{data}") if data.length > 0
+      resp = case pdu.app_tag
+      when Net::LDAP::PDU::BindRequest # bind request
+        cli.authenticated = true
+        encode_ldap_response(
+          pdu.message_id,
+          Net::LDAP::ResultCodeSuccess,
+          '',
+          '',
+          Net::LDAP::PDU::BindResult
+        )
+      when Net::LDAP::PDU::SearchRequest # search request
+        if cli.authenticated
+          # Perform query against some loaded LDIF structure
+          treebase = pdu.search_parameters[:base_object].to_s
+          # ... search, build packet, send to client
+          encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeNoSuchObject, "", "No such object", 5)
+        else
+          encode_ldap_response(pdu.message_id, Net::LDAP::ResultCodeUnwillingToPerform, "", "I'm sorry Dave, I can't do that", 5)
+        end
+      else
+        nil
+      end
+      resp.nil? ? cli.close : send_response(cli, resp)
+    rescue => e
+      elog(e)
+      cli.close
+      raise e
+    end
+  end
+
+  #
+  # Encode response for LDAP client consumption
+  #
+  # @param msgid [Integer] LDAP message identifier
+  # @param code  [Integer] LDAP message code
+  # @param dn    [String]  LDAP distinguished name
+  # @param msg   [String]  LDAP response message
+  # @param tag   [Integer] LDAP response tag
+  #
+  # @return [Net::BER::BerIdentifiedOid] LDAP query response
+  def encode_ldap_response(msgid, code, dn, msg, tag)
+    [
+      msgid.to_ber,
+      [
+        code.to_ber_enumerated,
+        dn.to_ber,
+        msg.to_ber
+      ].to_ber_appsequence(tag)
+    ].to_ber_sequence
+  end
+
+  #
+  # Search provided ldif data for query information
+  #
+  # @param filter [Net::LDAP::Filter] LDAP query filter
+  # @param attrflt [Array, Symbol] LDAP attribute filter
+  #
+  # @return [Array] Query matches
+  def search_ldif(filter, msgid, attrflt = :all)
+    return [] if @ldif.nil? or @ldif.empty?
+    ldif.map do |dn, entry|
+      if filter.match(entry)
+        attrs = []
+        entry.each do |k, v|
+          if attrflt == :all || attrflt.include?(k.downcase)
+            attrvals = v.map(&:to_ber).to_ber_set
+            attrs << [k.to_ber, attrvals].to_ber_sequence
+          end
+        end
+        appseq = [
+          dn.to_ber,
+          attrs.to_ber_sequence
+        ].to_ber_appsequence(4)
```

Comment:
Can use the constant here for readability.
```suggestion
        ].to_ber_appsequence(Net::LDAP::PDU::SearchReturnedData)
```

Reply:
Done